src/rgw/rgw_sal.h

```diff
@@ -1360,24 +1364,58 @@ class Lifecycle {
 						       const std::string& cookie) = 0;
 };
 
+/**
+ * @brief Abstraction for a Notification configuration entry
```

Comment:
why do we need abstraction via a base class?
in any specific place in the code, we know the exact type that is being encoded/decoded from the store.
won't it be better to just templatize by the type?

Reply:
I tried that first.  You cannot templatize in a pure virtual class.